torch/csrc/autograd/grad_buffer.cpp

```diff
@@ -19,6 +19,9 @@ auto GradBuffer::addGrad(size_t pos, std::shared_ptr<Variable>&& var) -> void {
   if (!item.first) {
     buffer[pos] = std::make_pair<>(std::move(tensor), true);
   } else {
+    if (item.first->isSparse() && !tensor->isSparse()) {
```

Comment:
ah, yes, you're right. My original intent was to disallow mixing entirely and then I realized it was trivial to allow this particular case. I'll update the code to automatically upgrade the result to dense if necessary, regardless of the addition order.